### PR TITLE
Some fixes for static files disk

### PR DESCRIPTION
**Pull request by kssenii**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by kssenii** [comment]
```
ATTACH TABLE test_hits UUID 'dbeecc07-58f3-4e89-9bee-cc0758f34e89'  ( WatchID UInt64,  JavaEnable UInt8,  Title String,  GoodEvent Int16,  EventTime DateTime,  EventDate Date,  CounterID UInt32,  ClientIP UInt32,  ClientIP6 FixedString(16),  RegionID UInt32,  UserID UInt64,  CounterClass Int8,  OS UInt8,  UserAgent UInt8,  URL String,  Referer String,  URLDomain String,  RefererDomain String,  Refresh UInt8,  IsRobot UInt8,  RefererCategories Array(UInt16),  URLCategories Array(UInt16), URLRegions Array(UInt32),  RefererRegions Array(UInt32),  ResolutionWidth UInt16,  ResolutionHeight UInt16,  ResolutionDepth UInt8,  FlashMajor UInt8, FlashMinor UInt8,  FlashMinor2 String,  NetMajor UInt8,  NetMinor UInt8, UserAgentMajor UInt16,  UserAgentMinor FixedString(2),  CookieEnable UInt8, JavascriptEnable UInt8,  IsMobile UInt8,  MobilePhone UInt8,  MobilePhoneModel String,  Params String,  IPNetworkID UInt32,  TraficSourceID Int8, SearchEngineID UInt16,  SearchPhrase String,  AdvEngineID UInt8,  IsArtifical UInt8,  WindowClientWidth UInt16,  WindowClientHeight UInt16,  ClientTimeZone Int16,  ClientEventTime DateTime,  SilverlightVersion1 UInt8, SilverlightVersion2 UInt8,  SilverlightVersion3 UInt32,  SilverlightVersion4 UInt16,  PageCharset String,  CodeVersion UInt32,  IsLink UInt8,  IsDownload UInt8,  IsNotBounce UInt8,  FUniqID UInt64,  HID UInt32,  IsOldCounter UInt8, IsEvent UInt8,  IsParameter UInt8,  DontCountHits UInt8,  WithHash UInt8, HitColor FixedString(1),  UTCEventTime DateTime,  Age UInt8,  Sex UInt8,  Income UInt8,  Interests UInt16,  Robotness UInt8,  GeneralInterests Array(UInt16), RemoteIP UInt32,  RemoteIP6 FixedString(16),  WindowName Int32,  OpenerName Int32,  HistoryLength Int16,  BrowserLanguage FixedString(2),  BrowserCountry FixedString(2),  SocialNetwork String,  SocialAction String,  HTTPError UInt16, SendTiming Int32,  DNSTiming Int32,  ConnectTiming Int32,  ResponseStartTiming Int32,  ResponseEndTiming Int32,  FetchTiming Int32,  RedirectTiming Int32, DOMInteractiveTiming Int32,  DOMContentLoadedTiming Int32,  DOMCompleteTiming Int32,  LoadEventStartTiming Int32,  LoadEventEndTiming Int32, NSToDOMContentLoadedTiming Int32,  FirstPaintTiming Int32,  RedirectCount Int8, SocialSourceNetworkID UInt8,  SocialSourcePage String,  ParamPrice Int64, ParamOrderID String,  ParamCurrency FixedString(3),  ParamCurrencyID UInt16, GoalsReached Array(UInt32),  OpenstatServiceName String,  OpenstatCampaignID String,  OpenstatAdID String,  OpenstatSourceID String,  UTMSource String, UTMMedium String,  UTMCampaign String,  UTMContent String,  UTMTerm String, FromTag String,  HasGCLID UInt8,  RefererHash UInt64,  URLHash UInt64,  CLID UInt32,  YCLID UInt64,  ShareService String,  ShareURL String,  ShareTitle String,  ParsedParams Nested(Key1 String,  Key2 String, Key3 String, Key4 String, Key5 String,  ValueDouble Float64),  IslandID FixedString(16),  RequestNum UInt32,  RequestTry UInt8) ENGINE = MergeTree() PARTITION BY toYYYYMM(EventDate) ORDER BY (CounterID, EventDate, intHash32(UserID)) SAMPLE BY intHash32(UserID) SETTINGS  storage_policy='web';


<storage_configuration>
    <disks>
        <web>
            <type>web</type>
            <endpoint>https://clickhouse-datasets.s3.yandex.net/test-hits/</endpoint>
        </web>
    </disks>
    <policies>
        <web>
            <volumes>
                <main>
                    <disk>web</disk>
                </main>
            </volumes>
        </web>
    </policies>
</storage_configuration>
```

File: src/Disks/DiskWebServer.cpp [hunk 1]
```diff
@@ -148,181 +133,106 @@ class ReadBufferFromWebServer final : public ReadIndirectBufferFromRemoteFS<Read
...
 
 size_t DiskWebServer::getFileSize(const String & path) const
 {
-    File file;
-    if (!findFileInMetadata(path, file))
-        throw Exception(ErrorCodes::LOGICAL_ERROR, "File {} not found", path);
-    return file.size;
+    auto iter = files.find(path);
+    if (iter == files.end())
+        return false;
```
Comment (evillique) [inline src/Disks/DiskWebServer.cpp:213]: Should it return `0`? Maybe even throw exception?
```suggestion
        return 0;
```

**Review by evillique — APPROVED** [review]

File: src/Disks/DiskWebServer.cpp (hunk 1 above)
Comment (kssenii) [reply src/Disks/DiskWebServer.cpp:213]: ups, thanks!
we do need to throw here.

**Comment by kssenii** [comment]
@Mergifyio update

**Comment by alexey-milovidov** [comment]
> docker: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit.

Ok.

**Comment by alexey-milovidov** [comment]
@Mergifyio update

**Comment by alexey-milovidov** [comment]
See the comment
https://github.com/ClickHouse/ClickHouse/pull/28603#issuecomment-912830451

It contains one extra whitespace.
After I copy-pasted it blindly, I get the following exception message:

```
Cannot load disk metadata. Error: Poco::Exception. Code: 1000, e.code() = 0, Bad URI syntax: URI contains invalid characters (version 21.11.1.8057). (NETWORK_ERROR)
```

Maybe print URI in quotes, so the user will see what characters are invalid?

**Comment by alexey-milovidov** [comment]
If I run `SELECT sum(cityHash64(*)) FROM test_hits WHERE CounterID = 29740601` with 1000 threads, I get:

```
Code: 210. DB::Exception: Received from localhost:9000. DB::Exception: Unreachable url: https://clickhouse-datasets.s3.yandex.net/disk-with-static-files-tests/test-hits/store/1ae/1ae36516-d62d-4218-9ae3-6516d62da218/201403_7_12_1/DOMInteractiveTiming.bin. Error: Received error from remote server /disk-with-static-files-tests/test-hits/store/1ae/1ae36516-d62d-4218-9ae3-6516d62da218/201403_7_12_1/DOMInteractiveTiming.bin. HTTP status code: 429 Too Many Requests, body: <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>TooManyRequests</Code><Message>Reduce request rate.</Message><Resource>/disk-with-static-files-tests/test-hits/store/1ae/1ae36516-d62d-4218-9ae3-6516d62da218/201403_7_12_1/DOMInteractiveTiming.bin</Resource><RequestId>fe17479ea359e4ab</RequestId></Error>: (while reading column DOMInteractiveTiming): (while reading from part https://clickhouse-datasets.s3.yandex.net/disk-with-static-files-tests/test-hits/store/1ae/1ae36516-d62d-4218-9ae3-6516d62da218/201403_7_12_1/ from mark 196 with max_rows_to_read = 8192): While executing MergeTreeThread. (NETWORK_ERROR)
```

Need to add retries with exponential backoff (maybe in the next PR).

**Comment by alexey-milovidov** [comment]
For `SELECT sum(cityHash64(*)) FROM test_hits` is happens even with default number of threads.

**Comment by alexey-milovidov** [comment]
```
Code: 210. DB::Exception: Received from localhost:9000. DB::Exception: Unreachable url: https://datasets.clickhouse.tech/disk-with-static-files-tests/test-hits/store/1ae/1ae36516-d62d-4218-9ae3-6516d62da218/201403_19_24_1/GoodEvent.bin. Error: Net Exception: (while reading column GoodEvent): (while reading from part https://datasets.clickhouse.tech/disk-with-static-files-tests/test-hits/store/1ae/1ae36516-d62d-4218-9ae3-6516d62da218/201403_19_24_1/ from mark 0 with max_rows_to_read = 8192): While executing MergeTreeInOrder. (NETWORK_ERROR)
```

What "Net Exception"? Looks like the reason is lost.

**Comment by alexey-milovidov** [comment]
After I replaced `clickhouse-datasets.s3.yandex.net` to `datasets.clickhouse.tech` and after some warmup it started to work better.

**Comment by kssenii** [comment]
btw as I tested -- `DB::Exception: Unreachable url` can be removed if you increase these 3 settings values `http_connection_timeout, http_receive_timeout, http_receive_timeout` equally big
https://github.com/ClickHouse/ClickHouse/blob/e52f5fdfce95d5c73b2e1b8d947855686b6bbaf3/src/Disks/DiskWebServer.h#L49
In this pr I increased a default minimum for them for this disk but not much.

**Comment by kssenii** [comment]
> Need to add retries with exponential backoff

Added.

**Comment by alexey-milovidov** [comment]
Experiment with DiskWeb on a server in Yandex Cloud with 40 threads:

Run query `SELECT sum(cityHash64(*)) FROM test_hits` several times.

1. datasets.clickhouse.tech:
26.697 sec, 15.534 sec, 14.133 sec

2. clickhouse-datasets.s3.yandex.net:
Failed due to rate limit: https://markify.dev/p/E1ybmsq/text

File: src/Disks/DiskWebServer.cpp [hunk 2]
```diff
@@ -125,204 +112,149 @@ class ReadBufferFromWebServer final : public ReadIndirectBufferFromRemoteFS<Read
...
+        try
+        {
+            initialize(fs::path(url) / fs::path(path).parent_path());
+            return files.find(path) != files.end();
+        }
+        catch (...)
+        {
+            const auto message = getCurrentExceptionMessage(false);
+            bool can_throw = CurrentThread::isInitialized() && CurrentThread::get().getQueryContext();
+            if (can_throw)
```
Comment (azat) [inline src/Disks/DiskWebServer.cpp:167]: @kssenii why it cannot throw if it is loaded from the server context? (read server start) This makes the table simply empty, which is not good:

```
2022.10.28 15:18:30.739698 [ 10663 ] {} <Error> ReadWriteBufferFromHTTP: HTTP request to `http://127.0.0.1:8080/store/a3e/a3e65e1f-5fd4-47ed-9dbd-307f2586b52d/detached/.index` failed at try 1/1 with bytes read: 0/unknown. Error: Connection refused. (Current backoff wait is 100/1600 ms)
2022.10.28 15:18:30.841210 [ 10663 ] {} <Trace> DiskWeb: Cannot load disk metadata. Error: Poco::Exception. Code: 1000, e.code() = 111, Connection refused (version 22.11.1.1                                                       
```

```sql
SELECT *
FROM data_from_web

Query id: ee544a5e-3c67-4fb4-8f14-f8e4a082b237

Ok.

0 rows in set. Elapsed: 0.019 sec. 
```

Comment (alexey-milovidov) [reply src/Disks/DiskWebServer.cpp:167]: @azat this is clearly a wrong behavior, and should be fixed.

Comment (azat) [reply src/Disks/DiskWebServer.cpp:167]: https://github.com/ClickHouse/ClickHouse/pull/42800

Comment (kssenii) [reply src/Disks/DiskWebServer.cpp:167]: This was done in order to allow disk Web to start even if there is no connection on server startup.

**Pull request merged**